File: android/app/src/main/java/com/focusguard/e2e/E2EBootstrap.kt
package com.focusguard.e2e

/** Applies Detox launchArgs to MMKV and permission mocks before React Native starts. */
object E2EBootstrap {
    private data class PresetFlags(
        val resetStorage: Boolean,
        val skipOnboarding: Boolean,
        val permissionsGranted: Boolean,
    )

    private val presets =
        mapOf(
            "fresh" to PresetFlags(resetStorage = true, skipOnboarding = false, permissionsGranted = false),
            "onboarding" to PresetFlags(resetStorage = true, skipOnboarding = false, permissionsGranted = false),
            "permissions" to PresetFlags(resetStorage = true, skipOnboarding = true, permissionsGranted = false),
            "dashboard" to PresetFlags(resetStorage = true, skipOnboarding = true, permissionsGranted = true),
        )

    fun applyFromCachedLaunchArgs() {
        if (!E2EFeature.isEnabled()) {
            return
        }

        val presetFlags = E2ELaunchArgs.get("e2ePreset")?.let { presets[it] }
        if (presetFlags != null) {
            apply(presetFlags)
            return
        }

        val resetStorage = E2ELaunchArgs.get("e2eResetStorage") == "true"
        val skipOnboarding = E2ELaunchArgs.get("e2eSkipOnboarding") == "true"
        val permissionsGranted = E2ELaunchArgs.get("e2ePermissionsGranted") == "true"

        if (!resetStorage && !skipOnboarding && !permissionsGranted) {
            return
        }

        apply(
            PresetFlags(
                resetStorage = resetStorage,
                skipOnboarding = skipOnboarding,
                permissionsGranted = permissionsGranted,
            ),
        )
    }

    private fun apply(flags: PresetFlags) {
        if (flags.resetStorage) {
            E2EAppState.resetStorage()
        }

        if (flags.skipOnboarding) {
            E2EAppState.setOnboardingComplete()
        }

        E2EPermissionOverride.permissionsGranted = flags.permissionsGranted
    }
}

File: source/testing/e2eBootstrap.ts
/** @format */

import { configureE2EBootstrap, getE2ELaunchArg, isE2EEnabled } from '@/specs';

const parseBool = (value: string | null): boolean => value === 'true';

/**
 * Applies Detox launchArgs before React mounts.
 * Called from index.js so MMKV is seeded before Zustand hydration.
 */
export const applyE2EBootstrapFromLaunchArgs = (): void => {
  if (!isE2EEnabled()) {
    return;
  }

  const preset = getE2ELaunchArg('e2ePreset') as E2ELaunchPreset | null;
  if (preset && preset in E2E_LAUNCH_ARGS) {
    applyLaunchFlags(E2E_LAUNCH_ARGS[preset]);
    return;
  }

  applyLaunchFlags({
    e2eResetStorage: getE2ELaunchArg('e2eResetStorage'),
    e2eSkipOnboarding: getE2ELaunchArg('e2eSkipOnboarding'),
    e2ePermissionsGranted: getE2ELaunchArg('e2ePermissionsGranted'),
  });
};

const applyLaunchFlags = (flags: Record<string, string | null | undefined>): void => {
  const resetStorage = parseBool(flags.e2eResetStorage ?? null);
  const skipOnboarding = parseBool(flags.e2eSkipOnboarding ?? null);
  const permissionsGranted = parseBool(flags.e2ePermissionsGranted ?? null);

  if (!resetStorage && !skipOnboarding && !permissionsGranted) {
    return;
  }

  configureE2EBootstrap(skipOnboarding, permissionsGranted, resetStorage);
};

export type E2ELaunchPreset = 'fresh' | 'onboarding' | 'permissions' | 'dashboard';

export const E2E_LAUNCH_ARGS: Record<E2ELaunchPreset, Record<string, string>> = {
  fresh: {
    e2eResetStorage: 'true',
  },
  onboarding: {
    e2eResetStorage: 'true',
  },
  permissions: {
    e2eResetStorage: 'true',
    e2eSkipOnboarding: 'true',
  },
  dashboard: {
    e2eResetStorage: 'true',
    e2eSkipOnboarding: 'true',
    e2ePermissionsGranted: 'true',
  },
};
